Cleaning up the C code

xavierleroy opened the pull request:
This PR is the result of running the C code of OCaml through Clang 6.0 and GCC 8 with full warnings on.  This produces a lot of warnings!  Some of the warnings point to areas of the code that can obviously be improved or cleaned up.  That's what this PR does. 

It's better reviewed commit by commit.  The only nontrivial commit is https://github.com/ocaml/ocaml/pull/1812/commits/03d8f0caeabf19452683d64e9f63c845f2011c81, which replaces some uses of `int` by more appropriate types like `intnat`.  On a 64-bit platform, `int` is only 32 bits and may not represent correctly e.g. the length of a string or the size of an OCaml heap block.  This commit replaces a number of uses of `int` by other types that are 64-bit wide on 64-bit architectures, such as `intnat` or `uintnat` or `size_t` or `mlsize_t`.
    
Sometimes an `intnat` was used as an `int` and is intended as a Boolean (0 or 1); then it was replaced by an `int`.
    
There are many remaining cases where we assign a 64-bit quantity to a 32-bit `int` variable.  Either I believe these cases are safe (e.g. the 64-bit quantity is the difference between two pointers within an I/O buffer, something that always fits in 32 bits), or the code change was not obvious and too risky.

xavierleroy committed:
C cleanup: old-style, unprototyped function declarations

It's `int f(void)`, not `int f()`.  [-Wstrict-prototypes]

xavierleroy committed:
C cleanup: be more explicit about conversions involvig `float` and `double`

byterun/bigarray.c, byterun/ints.c:
  add explicit casts to clarify the intent
  renamed float field of conversion union from `d` to `f`.

byterun/compact.c, byterun/gc_ctrl.c:
  some local variables were of type `float` while all FP computations
  here are done in double precision;
  turned these variables into `double`.

[-Wdouble-promotion -Wfloat-conversion]

xavierleroy committed:
C cleanup: add explicit initialization of struct field `compare_ext`

[-Wmissing-field-initializers]

xavierleroy committed:
C cleanup: declare more functions "noreturn"

[-Wmissing-noreturn]

xavierleroy committed:
C cleanup: make CAMLassert compliant with ISO C

In `e1 ? e2 : e3`, expressions `e2` and `e3` must have the same type.
`e2` of type `void` and `e3` of type `int`, as in the original code,
is a GNU extension.

xavierleroy committed:
C cleanup: remove or conditionalize unused macros

Some macros were defined and never used.
Some other macros were always defined but conditionally used.

[-Wunused-macros]

xavierleroy committed:
C cleanup: replace some uses of `int` by more appropriate types like `intnat`

On a 64-bit platform, `int` is only 32 bits and may not represent correctly
the length of a string or the size of an OCaml heap block.

This commit replaces a number of uses of `int` by other types that
are 64-bit wide on 64-bit architectures, such as `intnat` or `uintnat`
or `size_t` or `mlsize_t`.

Sometimes an `intnat` was used as an `int` and is intended as a Boolean
(0 or 1); then it was replaced by an `int`.

There are many remaining cases where we assign a 64-bit quantity to a
32-bit `int` variable.  Either I believe these cases are safe
(e.g. the 64-bit quantity is the difference between two pointers
within an I/O buffer, something that always fits in 32 bits), or
the code change was not obvious and too risky.

[-Wshorten-64-to-32]

xavierleroy committed:
C cleanup: put `inline` before return type

`static inline void f(void)` is cleaner than `static void inline f(void)`.

[-Wold-style-declaration]

xavierleroy committed:
C cleanup: unused assignment to unused parameter

Looks very useless.  [-Wunused-but-set-parameter]

pmetzger commented:
I obviously am not as thorough as the compiler, but reading through these everything looked pretty good.

Unexpectedly, I learned something new reading these. I had been unaware that re-interpreting a bit pattern as a different type through the use of a union was legal, but I checked and it is indeed! (For those who are interested, the place in the C11 standard section 6.5.2.3, footnote 95.) I shall have to remember this going forward.

xavierleroy commented:
CI "precheck" passes.  The Travis failure is just the missing Changes entry.

xavierleroy commented:
> I had been unaware that re-interpreting a bit pattern as a different type through the use of a union was legal, but I checked and it is indeed! 

Yes, this practice ("type punning" using a union)  is standardized in C99 and in C11.  In C90 it was implementation-defined behavior.  The GCC manual has always advertised it as safe even in the presence of strict type-based nonaliasing analysis.

mshinwell commented:
I will review this within the next few days.

let-def commented on byterun/caml/sys.h:
```diff
@@ -26,8 +26,14 @@ extern "C" {
+
 CAMLextern double caml_sys_time_unboxed(value);
 CAMLextern void caml_sys_init (char_os * exe_name, char_os ** argv);
 CAMLextern value caml_sys_exit (value);
```
Should `caml_sys_exit` be considered noreturn too?

xavierleroy replied:
It should, except that if the "C plugins" mechanism is active (`CAML_WITH_CPLUGINS` is defined), there is no guarantee that `caml_sys_exit` will not return.

We could add more #ifdefs to declare `caml_sys_exit` specially if `CAML_WITH_CPLUGINS` is not defined.  We could also get rid of this "C plugins" instrumentation hell.

avsm replied:
So the warnings from the static analyser will still strike if we supply an OCaml build with `CAML_WITH_CPLUGINS` defined :-/  We're very likely to be compiling MirageOS with the plugins option disabled soon since we trust it about as far as we can throw a small stick.

xavierleroy replied:
`CAML_WITH_CPLUGINS` is no longer defined by default.  Since 4.06, you need to `configure -with-cplugins` to activate it.  Still, at some future point we should review the actual uses of this mechanism and decide whether to keep it or not.

let-def commented on otherlibs/raw_spacetime_lib/spacetime_offline.c:
```diff
@@ -83,7 +83,6 @@ CAMLprim value caml_spacetime_unmarshal_trie (value v_channel)
 
 CAMLprim value caml_spacetime_node_num_header_words(value unit)
 {
-  unit = Val_unit;
```
Isn't
```c
  (void)unit;
```
the usual way to bypass unused warnings?

xavierleroy replied:
We don't try to bypass "unused argument" warnings, because 1- we have a lot of functions with legitimately-unused arguments, 2- we don't activate the "unused argument" warning, and 3- if we were to activate it, I'd rather mark unused arguments using GCC/Clang attributes.

let-def reviewed (changes requested):
The changes look good to me.

damiendoligez commented on byterun/ints.c:
```diff
@@ -286,13 +286,13 @@ static int32_t caml_swap32(int32_t x)
 
 value caml_int32_direct_bswap(value v)
-{ return caml_swap32(v); }
+{ return caml_swap32((int32_t) v); }
```
What about the implicit cast on return? `caml_swap32` returns a `int32_t`, not a `value`.

xavierleroy replied:
Indeed I only looked at 64->32 bit conversions -- those that lose information and are better made explicit.  int32_t to value doesn't lose information, so perhaps it can remain implicit.  I don't feel strongly about it either way.

damiendoligez replied:
The asymmetry just feels slightly odd, but it can stay this way.

damiendoligez commented on byterun/major_gc.c:
```diff
@@ -231,7 +231,7 @@ static void init_sweep_phase(void)
 
 /* auxillary function of mark_slice */
-static inline value* mark_slice_darken(value *gray_vals_ptr, value v, int i,
+static inline value* mark_slice_darken(value *gray_vals_ptr, value v, intnat i,
```
`i` is a field offset, so it should probably be `mlsize_t` (and that's what gets passed for this parameter in `mark_slice`).

xavierleroy replied:
OK, commit fd3fd43 uses `mlsize_t` as suggested.

damiendoligez reviewed (approved):
LGTM with two small comments.

xavierleroy committed:
C cleanup: replace an "intnat_t" by a "mlsize_t"

@damiendoligez prefers it this way.

shindere commented:
Xavier Leroy (2018/06/04 07:55 -0700):
> xavierleroy commented on this pull request.
> 
> 
> 
> > @@ -26,8 +26,14 @@ extern "C" {
>  
>  #define NO_ARG Val_int(0)
>  
> -CAMLextern void caml_sys_error (value);
> -CAMLextern void caml_sys_io_error (value);
> +CAMLnoreturn_start
> +CAMLextern void caml_sys_error (value)
> +CAMLnoreturn_end;
> +
> +CAMLnoreturn_start
> +CAMLextern void caml_sys_io_error (value)
> +CAMLnoreturn_end;
> +
>  CAMLextern double caml_sys_time_unboxed(value);
>  CAMLextern void caml_sys_init (char_os * exe_name, char_os ** argv);
>  CAMLextern value caml_sys_exit (value);
> 
> It should, except that if the "C plugins" mechanism is active (`CAML_WITH_CPLUGINS` is defined), there is no guarantee that `caml_sys_exit` will not return.
> 
> We could add more #ifdefs to declare `caml_sys_exit` specially if
> `CAML_WITH_CPLUGINS` is not defined.  We could also get rid of this "C
> plugins" instrumentation hell.

I'd say that any of the two proposed solutions looks good to me and, if
I may say, nicer than the current situation.

shindere commented:
Xavier Leroy (2018/06/04 14:57 +0000):
> xavierleroy commented on this pull request.
> 
> 
> 
> > @@ -83,7 +83,6 @@ CAMLprim value caml_spacetime_unmarshal_trie (value v_channel)
>  
>  CAMLprim value caml_spacetime_node_num_header_words(value unit)
>  {
> -  unit = Val_unit;
> 
> We don't try to bypass "unused argument" warnings, because 1- we have
> a lot of functions with legitimately-unused arguments, 2- we don't
> activate the "unused argument" warning, and 3- if we were to activate
> it, I'd rather mark unused arguments using GCC/Clang attributes.

If there is a way to tell the compiler one knows an argument is really
unused, wouldn't it be worth it activating this warning?

shindere commented:
Xavier Leroy (2018/06/04 10:15 -0700):
> xavierleroy commented on this pull request.
> 
> 
> 
> > @@ -26,8 +26,14 @@ extern "C" {
>  
>  #define NO_ARG Val_int(0)
>  
> -CAMLextern void caml_sys_error (value);
> -CAMLextern void caml_sys_io_error (value);
> +CAMLnoreturn_start
> +CAMLextern void caml_sys_error (value)
> +CAMLnoreturn_end;
> +
> +CAMLnoreturn_start
> +CAMLextern void caml_sys_io_error (value)
> +CAMLnoreturn_end;
> +
>  CAMLextern double caml_sys_time_unboxed(value);
>  CAMLextern void caml_sys_init (char_os * exe_name, char_os ** argv);
>  CAMLextern value caml_sys_exit (value);
> 
> `CAML_WITH_CPLUGINS` is no longer defined by default.  Since 4.06, you
> need to `configure -with-cplugins` to activate it.  Still, at some
> future point we should review the actual uses of this mechanism and
> decide whether to keep it or not.

Perhaps, meanwhile, it would be good to have the idfed and to mark the
funciton as noreturn when the feature is disabled, since this seems to
be not only the default situation but also the most frequent one.

shindere commented:
Just a general question: would we gain something and is it doable to try to
use types with a "semantics"? I mean, a type whos ename would describe more
precisely the kind of value that is stored? I am thinking for instance about
intnat which I have to say does not mean a lot to me. But perhaps this is
already the case and then it's just a matter of documenting it somewhere?

damiendoligez commented:
In fact we would probably gain slightly in readability by removing most of the types that are aliases to `uintnat` (aka `value`) but it's not likely to be worth the amount of work and the risk of introducing bugs.

xavierleroy committed:
C cleanup: add "noreturn" declaration for caml_sys_exit

But not if CAML_WITH_CPLUGINS is defined, because a plugin
could cause caml_sys_exit to return.

xavierleroy commented:
@let-def and @shindere: your wish came true, I added some "noreturn" declarations for `caml_sys_exit` but only if CPLUGINS is not active.  Commit https://github.com/ocaml/ocaml/pull/1812/commits/56eab46189103c59384ba07f03035c6654cf7a70 .  Now I'm running CI precheck again to make sure it doesn't break the Windows ports.

shindere commented:
Many thanks, @xavierleroy!

murmour commented:
Relevant: what's the point of `asize_t`?
```ocaml
typedef size_t asize_t;
```

Is this a historical artefact?

Both `size_t` and `asize_t` are used in the runtime, but perhaps it'd be better to replace all uses of `asize_t` with the standard `size_t`? If the change makes sense, I can do it (literally a 5-minute job).

Edit: grammar.

murmour commented on byterun/sys.c:
```diff
@@ -82,8 +82,8 @@ CAMLexport void caml_sys_error(value arg)
   } else {
-    int err_len = strlen(err);
-    int arg_len = caml_string_length(arg);
+    mlsize_t err_len = strlen(err);
```
Shouldn't this be `size_t`?

murmour replied:
Disregard that. I'm a crazy person who hates implicit conversions, compiles code with `-Wconversion`, and inserts tons of carefully thought-out casts.

Perhaps doing the same for camlrun would help catch a couple of bugs related to integer overflow or lossy integer undercasts, but that is a huge undertaking, and many people would hate the abundance of casts (some of my colleagues do).

murmour commented on byterun/startup.c:
```diff
@@ -508,7 +508,7 @@ CAMLexport value caml_startup_code_exn(
   caml_init_debug_info();
   if (caml_debugger_in_use) {
-    int len, i;
+    uintnat len, i;
```
Both should be `size_t`.

murmour commented:
Typo in one of commit messages: "involvig" -> "involving".

murmour commented:
I've reviewed the PR, and to me it looks all very good.

xavierleroy commented:
@mshinwell : thanks for you kind offer to review, but we already have 5 reviews and 30 comments for a fairly trivial PR, so don't feel obliged to add your own...

mshinwell commented:
@xavierleroy Yes indeed it looks like this has had sufficient attention, so I don't think I need to read this in detail.

xavierleroy commented:
@murmour: yes, type `asize_t` is a survivance from the Caml Light days, where a few crummy C compilers and libraries would not define `size_t`.  We can remove `asize_t` but I'd feel safer if we can make sure that no Caml/C bindings in OPAM packages use `asize_t`.

xavierleroy commented:
@shindere @murmour @damiendoligez: indeed, there is a tension between
* defining more type aliases with an intended meaning (note we already have quite a few: `value` for encoded integers or heap pointers, `header_t` for block headers, `mlsize_t` for OCaml heap block sizes and field indices, etc);
* defining fewer of them because they try to make a distinction that makes no sense (e.g. `mlsize_t` could just as well be `size_t`).

I don't think we will ever agree on where to put the cursor.  

What we can do is consider specific cases (of adding a new type alias, or removing an existing one), but even this may be a lot of arguing for little benefits.

murmour commented:
> We can remove asize_t but I'd feel safer if we can make sure that no Caml/C bindings in OPAM packages use `asize_t.

A safe option is to replace _uses_ of `asize_t` with `size_t`, but keep the definition of `asize_t`, marking it as deprecated.

xavierleroy commented:
Thanks, all, for the feedback.  It's merged!
